### PR TITLE
Add stub for new goto-harness tool [depends-on: #3920]

### PR DESCRIPTION
**Pull request by hannes-steffenhagen-diffblue**
This adds a new executable called `goto-harness`. Right now it doesn't actually do anything, but ultimately its purpose will be to generate "harness" functions for goto-programs - i.e. given some specification, it'll generate a function suitable as a cbmc entry point function that implements that specification.

Planned for now are two types of harnesses:

One that takes the name of a function with parameters and then generates a function that sets up these parameters and calls the function with them. This is similar to what `cbmc` does already, however will allow more flexibility in choosing how exactly these parameters will be initialised which was deemed out of scope for `cbmc`.

The other will be able to take a snapshot of a concrete execution of a program, and then start an analysis from that point.

This PR should serve as a common start point for both of these pieces of work. It's marked as **Work In Progress** for now as we've not settled on a sensible command line interface; The problem being that different harness generators have mostly completely different sets of options, with only a handful of common options. 

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/Makefile [hunk 1]
```diff
@@ -27,6 +28,7 @@ all: cbmc.dir \
      goto-cc.dir \
      goto-diff.dir \
      goto-instrument.dir \
+		 goto-harness.dir \
```
Comment (tautschnig) [inline src/Makefile]: Wrong indent or tabs vs spaces issue.

File: src/goto-harness/goto_harness_parse_options.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    auto const fill_left = fill / 2;
+    auto const fill_right = fill - fill_left;
+    return border + std::string(fill_left, ' ') + text +
+           std::string(fill_right, ' ') + border;
+  };
+  std::cout << '\n'
+            << banner_string("Goto-Harness", CBMC_VERSION) << '\n'
+            << align_center_with_border("Copyright (C) 2019") << '\n'
+            << align_center_with_border("Diffblue Ltd.") << '\n'
+            << align_center_with_border("info@diffblue.com")
```
Comment (tautschnig) [inline src/goto-harness/goto_harness_parse_options.cpp:36]: We should likely have this `align_center_with_border` code for everyone...

**Review by tautschnig — COMMENTED** [review]
1) Please copy some of the description into the commit message.
2) It would likely be good to include a regression test directory (without any actual tests just yet, just the makefiles) in the initial commit as well.

With those we'd in fact have a very nice commit to serve as template for further tools to be created.

**Review by sonodtt — APPROVED** [review]
Looks good! Once again, many thanks for the introduction you give. Awesome. 

Very interesting the codesign section of the makefile :)

File: src/goto-harness/goto_harness_parse_options.cpp (hunk 2 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-harness/goto_harness_parse_options.cpp:36]: https://github.com/diffblue/cbmc/pull/3920

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: ff0da3a2f7ff7fd50829e580d73a6433ceb27236).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98638933
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by hannes-steffenhagen-diffblue** [comment]
test-gain failures are non-transient (i.e. not going to magically disappear with a force push) but unrelated to this PR

**Review by tautschnig — APPROVED** [review]
Repeating my earlier wishlist request: please also add a regression test folder and quite possibly a first regression test, simply running --version (and testing the exit code). That will make this commit a very nice template for future tools.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: a07b2cbbefc160f27999d429a55548ffa08e598c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98650165
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by hannes-steffenhagen-diffblue** [comment]
@tautschnig Ah sorry, forgot about that! Yeah that makes sense, I've added this now! (I also checked that the regression test is indeed executed - do you know if there is a good way to ensure this mechanically? The only method I know is deliberately failing a test; Accidentally adding adding tests without making sure they are actually being run has been a problem in the past)

__EDIT__  I forgot to do the makefile version of the regression, adding that now

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 9459b3022513c6d04bf11adab32776e95fed57ef).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98652328
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 30cb51c22affc07a212611b5c6420c127933b8d4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98652084
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 4064bc7ac6a480b4ae324d45d446d7a8cbda5223).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98656522
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 7a8d3a8b5ac09e00bbefa5648a58dbecd15278a7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98658100
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Pull request merged**